### PR TITLE
Added support for multiple objectmappers for different mediatypes

### PR DESCRIPTION
**Pull request by janbj**

### CODE REVIEW

File: apigateway/src/main/java/nva/commons/apigateway/ApiGatewayHandler.java [hunk 1]
```diff
@@ -62,6 +67,19 @@ public void init(OutputStream outputStream, Context context) {
...
 
+    /**
+     * Get the ObjectMapper to use for the given MediaType. Defaults to defaultRestObjectMapper if no other
+     * ObjectMapper is found.
+     *
+     * @param requestInfo   the requestInfo object with information about MediaType
+     * @return objectMapper the objectMapper to use for this MediaType
+     * @throws UnsupportedAcceptHeaderException when Accept Header value is unsupported by the service
+     */
+    public ObjectMapper getObjectMapper(RequestInfo requestInfo) throws UnsupportedAcceptHeaderException {
```
Comment (axthosarouris) [inline apigateway/src/main/java/nva/commons/apigateway/ApiGatewayHandler.java:78]: Do we have an example for this use case?  The included test uses the default objectMapper and not one of the Map of the mediaTypes to objectMapper.

File: apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java [hunk 2]
```diff
@@ -359,6 +361,21 @@ void handlerSendsRedirectionWhenItReceivesARedirectException() throws IOExceptio
         assertThat(response.getHeaders(), hasEntry(HttpHeaders.LOCATION, expectedRedirectLocation.toString()));
     }
 
+    @Test
+    void handlerSendsRawJsonWhenUsingPassthroughObjectMapper() throws Exception {
```
Comment (axthosarouris) [inline apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java:365]: Can't this be solved by the handler returning JsonNode?  i.e. 

```
public RawJsonResponseHandler(){


public JsonNode processInput(){

 return Try.attempt(() -> JsonUtils.dtoObjectMapper.writeValueAsString(body))
                   .map(bodyString->JsonUtils.dtoObjectMapper.readTree(bodyString)).orElseThrow();
}

}

File: buildSrc/src/main/groovy/nvacommons.java-conventions.gradle [hunk 3]
```diff
@@ -9,7 +9,7 @@ plugins {
 }
 
 group 'com.github.bibsysdev'
-version = '1.20.13-beta3'
+version = '1.20.13-beta4'
```
Comment (axthosarouris) [inline buildSrc/src/main/groovy/nvacommons.java-conventions.gradle:12]: Just put 1.20.14. I have created a new package apigatewayv2 that is not ready yet, but I do not think it is such a problem.

File: apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java (hunk 2 above)
Comment (janbj) [reply apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java:365]: We want to support XML in the response.

File: apigateway/src/main/java/nva/commons/apigateway/ApiGatewayHandler.java (hunk 1 above)
Comment (janbj) [reply apigateway/src/main/java/nva/commons/apigateway/ApiGatewayHandler.java:78]: I will create a test for this.

File: apigateway/src/main/java/nva/commons/apigateway/GatewayResponse.java [hunk 4]
```diff
@@ -63,30 +66,49 @@ public GatewayResponse(T body, Map<String, String> headers, int statusCode, Obje
      * Handler directly and we want to read the output.
      *
      * @param outputStream the outputStream updated by the lambda handler
-     * @param <T>          the class of the body
      * @return the GatewayResponse containing the output of the handler
      * @throws JsonProcessingException when the OutputStream cannot be parsed to a JSON object.
      */
-    public static <T> GatewayResponse<T> fromOutputStream(ByteArrayOutputStream outputStream)
+    public static <T> GatewayResponse<T> fromOutputStream(ByteArrayOutputStream outputStream, Class<T> className)
```
Comment (axthosarouris) [inline apigateway/src/main/java/nva/commons/apigateway/GatewayResponse.java:72]: Do not break the API. Keep the old the methods but deprecate them.

```
  * Create GatewayResponse object from an output stream. Used when we call the method {@code handleRequest()} of a
     * Handler directly and we want to read the output.
     *
     * @param outputStream the outputStream updated by the lambda handler
     * @return the GatewayResponse containing the output of the handler
     * @throws JsonProcessingException when the OutputStream cannot be parsed to a JSON object.
     */
    public static <T> GatewayResponse<T> fromOutputStream(ByteArrayOutputStream outputStream, Class<T> className)
        throws JsonProcessingException {
        String json = outputStream.toString(StandardCharsets.UTF_8);
        return fromString(json, className);
    }

    @Deprecated(forRemoval = true)
    public static <T> GatewayResponse<T> fromOutputStream(ByteArrayOutputStream outputStream)
        throws JsonProcessingException {
        String json = outputStream.toString(StandardCharsets.UTF_8);
        return fromString(json);
    }

    /**
     * Create GatewayResponse object from a String. Used when we call the method {@code handleRequest()} of a Handler
     * directly and we want to read the output. Usually the String is the output of an OutputStream.
     *
     * @param responseString a String containing the serialized GatwayResponse
     * @return the GatewayResponse containing the output of the handler
     * @throws JsonProcessingException when the OutputStream cannot be parsed to a JSON object.
     */
    public static <T> GatewayResponse<T> fromString(String responseString, Class<T> className)
        throws JsonProcessingException {
        if (responseBodyIsString(className)) {
            return constructGatewayResponseWithStringBody(responseString);
        } else {
            return constructResponseWithJsonObjectBody(responseString);
        }
    }

    @Deprecated(forRemoval = true)
    public static <T> GatewayResponse<T> fromString(String responseString)
        throws JsonProcessingException {
        return constructResponseWithJsonObjectBody(responseString);
    }

    private static <T> boolean responseBodyIsString(Class<T> className) {
        return className.getTypeName().equals(String.class.getTypeName());
    }

    private static <T> GatewayResponse<T> constructResponseWithJsonObjectBody(String responseString)
        throws JsonProcessingException {
        TypeReference<GatewayResponse<T>> typeref = new TypeReference<>() {
        };
        return defaultRestObjectMapper.readValue(responseString, typeref);
    }

    private static <T> GatewayResponse<T> constructGatewayResponseWithStringBody(String responseString)
        throws JsonProcessingException {
        JsonNode jsonNode = defaultRestObjectMapper.readTree(responseString);
        String body = jsonNode.get("body").asText();
        TypeReference<Map<String, String>> typeref = new TypeReference<>() {
        };
        Map<String, String> headers = defaultRestObjectMapper.convertValue(jsonNode.get("headers"), typeref);
        int statusCode = jsonNode.get("statusCode").asInt();
        return (GatewayResponse<T>)
                   attempt(() -> new GatewayResponse(body, headers, statusCode, defaultRestObjectMapper)).orElseThrow();
    }
```

File: apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java [hunk 5]
```diff
@@ -359,6 +365,34 @@ void handlerSendsRedirectionWhenItReceivesARedirectException() throws IOExceptio
         assertThat(response.getHeaders(), hasEntry(HttpHeaders.LOCATION, expectedRedirectLocation.toString()));
     }
 
+    @Test
+    void handlerRespondsWithJsonWhenClientAsksForJson() throws Exception {
```
Comment (axthosarouris) [inline apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java:369]: shouldRetrurnJsonObjectWhenClientAsksForJson

File: apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java [hunk 6]
```diff
@@ -359,6 +365,34 @@ void handlerSendsRedirectionWhenItReceivesARedirectException() throws IOExceptio
...
+        handler.handleRequest(inputStream, outputStream, context);
+
+        var response = GatewayResponse.fromOutputStream(outputStream, String.class);
+        var requestBody = dtoObjectMapper.readValue(response.getBody(), RequestBody.class);
+
+        assertThat(requestBody, is(notNullValue()));
+    }
+
+    @Test
+    void handlerRespondsWithXmlBodyWhenClientAsksForXml() throws Exception {
```
Comment (axthosarouris) [inline apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java:382]: ditto

File: apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java [hunk 7]
```diff
@@ -359,6 +365,34 @@ void handlerSendsRedirectionWhenItReceivesARedirectException() throws IOExceptio
...
+        var objectMapper = dtoObjectMapper;
```
Comment (axthosarouris) [inline apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java:383]: unusled line

File: apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java [hunk 8]
```diff
@@ -359,6 +365,34 @@ void handlerSendsRedirectionWhenItReceivesARedirectException() throws IOExceptio
...
+        var handler = new RawStringResponseHandler(dtoObjectMapper);
+        var inputStream = requestWithAcceptXmlHeader();
+        var outputStream = outputStream();
+        handler.handleRequest(inputStream, outputStream, context);
+
+        var response = GatewayResponse.fromOutputStream(outputStream, String.class);
+        var xmlMapper = new XmlMapper();
+        var requestBody = xmlMapper.readValue(response.getBody(), RequestBody.class);
+
+        assertThat(requestBody, is(notNullValue()));
```
Comment (axthosarouris) [inline apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java:393]: if the handler is echoing the request, can you verify that you are receiving the request obejct?

File: apigateway/src/main/java/nva/commons/apigateway/GatewayResponse.java (hunk 4 above)
Comment (axthosarouris) [reply apigateway/src/main/java/nva/commons/apigateway/GatewayResponse.java:72]: In the above comment, you can add @JacocoGenerated to deprecated methods

**Review by axthosarouris — CHANGES_REQUESTED** [review]

**Review by axthosarouris — DISMISSED** [review]

**Review by axthosarouris — APPROVED** [review]

**Pull request merged**